docs: extracting state logic into a reducer (#480)
reactjs/it.react.dev: src/content/learn/extracting-state-logic-into-a-reducer.md

---

> title: Estrarre la logica dello State in un Reducer

Reviewer: ```suggestion
title: Estrarre la Logica dello State in un Reducer
```

---

<Intro>

> I componenti con molti aggiornamenti di state distruibiti su molti event handler può diventare eccessivo. In questi casi, è possibile consolidare tutti gli aggiornamenti della logica dello state fuori dal componente in una singola funzione, chiamata _reducer._

Reviewer: ```suggestion

I componenti con molti aggiornamenti di state distruibiti su molti event handler può diventare eccessivo. In questi casi, è possibile consolidare tutti gli aggiornamenti della logica dello state fuori dal componente in una singola funzione, chiamata _reducer._
```

Reviewer: Teniamo lo stesso numero di linee rispetto all'originale :)

}
```

Rimuovi tutta la logica del setting state. Cosa devi lasciare con i tre event handler:

- `handleAddTask(text)` viene chiamato quando l'utente preme "Add".
- `handleChangeTask(task)` viene chiamato quando l'utente aziona un task o preme "Save".
- `handleDeleteTask(taskId)` viene chiamato quando l'utente preme "Delete".

> Gestire lo state con i reducer è leggermente diverso dall'utilizzare direttamente un setting state. Invece di dire a React "cosa fare" utilizzando setting state, specifichi "cosa lo user ha appena fatto" utilizzando delle azioni di dispatching che provengono dai tuoi event handler. (Lo state che aggiorna la logica vive da un altra parte!) Quindi invece di "setting `tasks`" tramite un event handler, stai utiilizzando un'azione di dispatching come "aggiungere/modificare/cancellare un task". Questo descrive molto di più l'intenzione dell'utente.

Reviewer: ```suggestion
Gestire lo state con i reducer è leggermente diverso dall'utilizzare direttamente un setting state. Invece di dire a React "cosa fare" utilizzando setting state, specifichi "cosa lo user ha appena fatto" utilizzando delle azioni di dispatching che provengono dai tuoi event handler. (Lo state che aggiorna la logica vive da un altra parte!) Quindi invece di "impostare `tasks`" tramite un event handler, stai utiilizzando un'azione di dispatching come "aggiungere/modificare/cancellare un task". Questo descrive molto di più l'intenzione dell'utente.
```

}
```

È un normale oggetto JavaScript. Decidi tu cosa metterci dentro, ma generalmente dovrebbe contenere la minima informazione riguardo _cosa è successo_. (Aggiungerai la funzione di `dispatch` in un altro passaggio.)

<Note>

Un oggetto action può avere qualsiasi forma.

> Per convenzione, è comune dare un `type` stringa che descriva cosa è accaduto e per passare altre informazioni aggiuntive agli altri campi. Il `type` è specifico ad un componente, dunque in questo esempio sia `'added'` o `'added_task'` vanno bene. Scegli un nome che descriva cosa è accaduto!

Reviewer: ```suggestion
Per convenzione, è comune dare un `type` stringa che descriva cosa è accaduto e per passare altre informazioni aggiuntive agli altri campi. Il `type` è specifico ad un componente, dunque in questo esempio sia `'added'` che `'added_task'` vanno bene. Scegli un nome che descriva cosa è accaduto!
```

```js
dispatch({
  // specifico del componente
  type: 'what_happened',
  // qua vanno gli altri campi
});
```

</Note>

### Step 2: Scrivi una funzione reducer {/*step-2-write-a-reducer-function*/}

Una funzione reducer è dove metterai la tua logica dello state. Prende due argomenti, lo state corrente e l'oggetto action e ritorna il nuovo state:

```js
function yourReducer(state, action) {
  // ritorna il nuovo state per React da utilizzare
}
```

React metterà quello che viene ritornato dalla funzione reducer nello state.
Per spostare la logica set dello state dai tuoi event handler in una funzione reducer in questo esempio, dovrai:

1. Dichiarare il state corrente (`tasks`) come primo argomento.
2. Dichiarare l'oggetto `action` come secondo argomento.
3. Ritornare il _next_ state dal reducer (il quale verrà collocato nello state da React).

> Qui c'è tutta la logica del set state migrata in una funzione reducer:

Reviewer: ```suggestion
Qui c'è tutta la logica di impostazione dello state migrata in una funzione reducer:
```

}
```

Poiché la funzione reducer prende lo state (`tasks`) come argomento, **puoi dichiararla all'esterno del tuo componente.** Ciò riduce il livello di indentazione e può rendere il tuo codice più leggibile.

<Note>

> Il codice sopra utilizza le istruzioni if/else, ma è una convenzione utilizzare le [istruzioni switch](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/switch) all'interno dei reducers. Il risultato è lo stesso, ma le istruzioni switch possono essere più facili da leggere a colpo d'occhio.

Reviewer: ```suggestion
Il codice sopra utilizza le istruzioni if/else, ma è una convenzione utilizzare le [istruzioni switch](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/switch) all'interno dei reducer. Il risultato è lo stesso, ma le istruzioni switch possono essere più facili da leggere a colpo d'occhio.
```

); // 1 + 2 + 3 + 4 + 5
```

> La funzione che passi al `reduce` è conosciuta come un "riduttore". Essa prende il _risultato fino a quel momento_ e l'_elemento corrente_, per poi restituire il _prossimo risultato._ I reducers di React sono un esempio della stessa idea: essi prendono _lo stato fino a quel momento_ e l'_azione_, restituendo poi lo _stato successivo_. In questo modo, essi accumulano le azioni nel tempo all'interno dello stato.

Reviewer: ```suggestion
La funzione che passi al `reduce` è conosciuta come un "reducer". Essa prende il _risultato fino a quel momento_ e l'_elemento corrente_, per poi restituire il _prossimo risultato._ I reducer di React sono un esempio della stessa idea: essi prendono _lo state fino a quel momento_ e l'_azione_, restituendo poi lo _state successivo_. In questo modo, essi accumulano le azioni nel tempo all'interno dello state.
```

</Sandpack>

Probabilmente non dovrai farlo da solo, ma questo è simile a ciò che fa React!

</DeepDive>

### Step 3: Usa il reducer dal tuo componente {/*step-3-use-the-reducer-from-your-component*/}

Finalmente, devi collegare il `tasksReducer` al tuo componente. Importa lo `useReducer` Hook da React:

```js
import { useReducer } from 'react';
```

Poi puoi sostituire `useState`:

```js
const [tasks, setTasks] = useState(initialTasks);
```

> con `useReducer` così:

Reviewer: ```suggestion
Con `useReducer` così:
```

</Sandpack>

Probabilmente non dovrai farlo da solo, ma questo è simile a ciò che fa React!

</DeepDive>

### Step 3: Usa il reducer dal tuo componente {/*step-3-use-the-reducer-from-your-component*/}

Infine, devi collegare il `tasksReducer` al tuo componente. Importa l'Hook `useReducer` da React:

```js
import { useReducer } from 'react';
```

Poi puoi sostituire `useState`:

```js
const [tasks, setTasks] = useState(initialTasks);
```

Con `useReducer` così:

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

> Lo `useReducer` Hook è simile allo `useState`-devi passargli uno stato iniziale e lui ti restituisce un valore dello stato e un modo per impostare lo stato (in questo caso, la funzione dispatch). Ma è un po' diverso.

Reviewer: ```suggestion
L'Hook `useReducer` è simile allo `useState`-devi passargli uno stato iniziale e lui ti restituisce un valore dello stato e un modo per impostare lo stato (in questo caso, la funzione dispatch). Ma è un po' diverso.
```

</Sandpack>

I reducers devono essere puri, quindi non dovrebbero mutare lo state. Ma Immer ti fornisce un oggetto `draft` speciale che è sicuro da mutare. Dietro le quinte, Immer creerà una copia del tuo state con le modifiche che hai apportato al `draft`. Questo è il motivo per cui i reducers gestiti da `useImmerReducer` possono mutare il loro primo argomento e non hanno bisogno di ritornare lo state.

<Recap>

- Per passare da `useState` a `useReducer`:

>   1. Esegui il dispatch delle azioni dagli event handlers.

Reviewer: ```suggestion
  1. Esegui il dispatch delle azioni dagli event handler.
```

</Solution>

#### Svuota l'input dopo l'invio di un messaggio {/*clear-the-input-on-sending-a-message*/}

Al momento, premere "Invia" non produce alcun effetto. Aggiungi un event handler al pulsante "Invia" che:

1. Mostrerà un `alert` con l'email del destinatario e il messaggio.
2. Svuoterà l'input del messaggio.

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js

Reviewer: ```suggestion
```js src/messengerReducer.js
```

}
```

> ```js ContactList.js

Reviewer: ```suggestion
```js src/ContactList.js
```

}
```

> ```js Chat.js active

Reviewer: ```suggestion
```js src/Chat.js active
```

<Solution>

> Ci sono un paio di modi che potresti utilizzare nel "Send" button event handler. Un approccio è quello di mostrare un alert e poi eseguire il dispatch di una azione `edited_message` con un `message` vuoto:

Reviewer: ```suggestion
Ci sono un paio di modi che potresti utilizzare nell'event handler del pulsante "Send". Un approccio è quello di mostrare un alert e poi eseguire il dispatch di una azione `edited_message` con un `message` vuoto:
```

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js

Reviewer: ```suggestion
```js src/messengerReducer.js
```

}
```

> ```js ContactList.js

Reviewer: ```suggestion
```js src/ContactList.js
```

}
```

> ```js Chat.js active

Reviewer: ```suggestion
```js src/Chat.js active
```

</Sandpack>

Questo metodo funziona e svuota l'input quando si preme "Invia".

Tuttavia, _dal punto di vista dell'utente_, inviare un messaggio è un'azione diversa rispetto alla modifica del campo. Per rappresentare questa distinzione, potresti invece creare una _nuova_ azione chiamata `sent_message`, e gestirla separatamente nel reducer:

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js active

Reviewer: ```suggestion
```js src/messengerReducer.js active
```

}
```

> ```js ContactList.js

Reviewer: ```suggestion
```js src/ContactList.js
```

}
```

> ```js Chat.js active

Reviewer: ```suggestion
```js src/Chat.js active
```

</Sandpack>

Il risultato finale è identico. Ma tieni presente che i tipi di azione dovrebbero idealmente descrivere "cosa ha fatto l'utente" piuttosto che "come vuoi che cambi lo state". Questo facilita l'aggiunta di ulteriori funzionalità in seguito.

> Con entrambe le soluzioni, è importante che tu **non** posizioni l'`alert` all'interno di un reducer. Il reducer dovrebbe essere una funzione pura - dovrebbe solo calcolare il prossimo state. Non dovrebbe "fare" nulla, compreso mostrare messaggi all'utente. Questo dovrebbe avvenire nel event handler. (Per aiutare a individuare errori come questo, React chiamerà i tuoi reducers più volte in modalità Strict. Questo è il motivo per cui, se inserisci un alert in un reducer, si attiva due volte.)

Reviewer: ```suggestion
Con entrambe le soluzioni, è importante che tu **non** posizioni l'`alert` all'interno di un reducer. Il reducer dovrebbe essere una funzione pura - dovrebbe solo calcolare il prossimo state. Non dovrebbe "fare" nulla, compreso mostrare messaggi all'utente. Questo dovrebbe avvenire nel event handler. (Per aiutare a individuare errori come questo, React chiamerà i tuoi reducer più volte in modalità Strict. Questo è il motivo per cui, se inserisci un alert in un reducer, si attiva due volte.)
```

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js

Reviewer: ```suggestion
```js src/messengerReducer.js
```

}
```

> ```js ContactList.js

Reviewer: ```suggestion
```js src/ContactList.js
```

}
```

> ```js Chat.js

Reviewer: ```suggestion
```js src/Chat.js
```

}
```

Dovresti anche aggiornare il componente `Messenger` per leggere il messaggio per il contatto attualmente selezionato:

```js
const message = state.messages[state.selectedId];
```

Questa è la soluzione completa:

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js

Reviewer: ```suggestion
```js src/messengerReducer.js
```

}
```

> ```js Chat.js hidden

Reviewer: ```suggestion
```js src/Chat.js hidden
```

<Solution>

L'esecuzione di un dispatch di una action chiama un reducer con lo state attuale e la action, e conserva il risultato come stato successivo. Ecco come si presenta nel codice:

<Sandpack>

> ```js App.js

Reviewer: ```suggestion
```js src/App.js
```

];
```

> ```js messengerReducer.js

Reviewer: ```suggestion
```js src/messengerReducer.js
```

}
```

> ```js MyReact.js active

Reviewer: ```suggestion
```js src/MyReact.js active
```

}
```

> ```js ContactList.js hidden

Reviewer: ```suggestion
```js src/ContactList.js hidden
```

</Sandpack>

Sebbene non faccia differenza nella maggior parte dei casi, un'implementazione leggermente più precisa si presenta così:

```js
function dispatch(action) {
  setState((s) => reducer(s, action));
}
```

> Questo perché le azioni inviate vengono messe in coda fino al prossimo rendering, [similmente alle funzioni di aggiornamento.](/learn/queueing-a-series-of-state-updates)

Reviewer: Alla fine del file c'è un mismatch di 4 righe (originale 2625, traduzione 2621), dovremmo cercare di tenere lo stesso numero di righe per evitare problemi durante le PR di sync
